Dispose scope

mfagerlund opened the pull request:
Suggested framework for managing disposing of tensors, using DisposeScopes.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=453) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=453) before we can accept your contribution.<br/><br/>:x: mfagerlund [sign now](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=453)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/TorchSharp?pullRequest=453) it.</sub>

NiklasGustafsson commented on src/TorchSharp/TorchSharp.csproj:
```diff
@@ -30,6 +30,7 @@
       <AutoGen>True</AutoGen>
       <DependentUpon>TensorTyped.tt</DependentUpon>
     </Compile>
+    <Compile Remove="Utils\torchu.cs" />
```
What is this about?

NiklasGustafsson commented on src/TorchSharp/Tensor/Tensor.cs:
```diff
@@ -27,11 +29,27 @@ public partial class Tensor : IDisposable
                 System.Threading.Interlocked.Increment(ref _totalCount);
                 _peakCount = Math.Max(_totalCount, _peakCount);
+
+                System.Threading.Interlocked.Add(ref _totalMemory, NumberOfElements * ElementSize);
```
This will result in overestimating the amount of memory. There are lots of ways that aliases for the same storage can be created. I actually had this in earlier, but took it out because it's not reliable.

mfagerlund replied:
When doing memory benchmarks, it's useful, but it could be placed in the DisposeScope if you prefer - I realized from your comments regarding PeakCount that it was always going to be aproximate. I'll remove it for now.

NiklasGustafsson commented on src/TorchSharp/Tensor/Tensor.cs:
```diff
@@ -1039,30 +1076,49 @@ public Tensor contiguous()
-                        (idx.kind == TensorIndex.Kind.Slice && !idx.startIndexOrBoolOrSingle.HasValue) ? long.MinValue + 6 :
-                        (idx.kind == TensorIndex.Kind.Single) ? idx.startIndexOrBoolOrSingle.GetValueOrDefault() :
-                        idx.startIndexOrBoolOrSingle.GetValueOrDefault() + long.MinValue / 2;
+                        (idx.kind == TensorIndex.Kind.Null)
```
Was this the result of some automatic re-formatting? I prefer the original. In general, I would suggest against PRs with reformatting, since it makes the diff much harder to read and spot the actual logical changes. Unless it's egregiously bad in the original, of course.

Same comment applies to all the re-formatting edits.

mfagerlund replied:
Fixed

NiklasGustafsson commented on src/TorchSharp/Tensor/Tensor.cs:
```diff
@@ -42,30 +60,35 @@ internal Tensor(IntPtr handle)
 
             public void Dispose()
             {
+                foreach (var callback in OnTensorDisposeCallback) {
```
I'm skeptical about this. I would like to better understand the performance impact of having this call on every tensor dispose.

mfagerlund replied:
There are the same number of calls when tensors are created, and deleting from a hash set is O(1), so I'm not sure it would be much of a problem?

NiklasGustafsson replied:
Yeah, as long as it's a constant-time operation, it's fine. The cost of removing items from a collection is always worrisome to me, and this is going to be on a lot of critical code paths, so I want to make sure we have it under control.

mfagerlund replied:
Let's keep an eye on performance and revisit this if we see any degradation?

NiklasGustafsson replied:
I'll run some benchmarks before releasing this to NuGet.

NiklasGustafsson commented on src/TorchSharp/DisposeScopeManager.cs:
```diff
@@ -0,0 +1,317 @@
+            private void DoDispose(IDisposable disposable)
+            {
+                _disposeScopeManager.CurrentlyDisposing = disposable;
+                if (disposable is torch.Tensor tensor) {
```
How about a try ... finally here to make sure that the 'CurrentlyDisposing' is always set to null after? I'm not sure that it's necessary, but it feels like one of those situations where I would expect it.

mfagerlund replied:
Fixed

NiklasGustafsson commented on test/TorchSharpTest/TestDisposeScopes.cs:
```diff
@@ -0,0 +1,188 @@
+            var preTotalCount = torch.Tensor.TotalCount;
+            using (var d = torch.NewDisposeScope()) {
+                var newValue = data * data + data;
+                Assert.True(torch.Tensor.TotalCount > preTotalCount);
```
As I mentioned in the issue #450 comments, since xUnit relies on multi-threading, this assertion is likely to fail intermittently. It's too bad, because it's a great way to test that the behavior is what you expect, but I speak from bitter experience.

mfagerlund replied:
I use the numbers coming from the DisposeScopeManager instead - it's thread static.

NiklasGustafsson commented on src/TorchSharp/Device.cs:
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
```
Seems to be nothing but reformatting.

mfagerlund replied:
I'll revert it.

NiklasGustafsson commented on src/TorchSharp/DisposeScopeManager.cs:
```diff
@@ -0,0 +1,317 @@
+using System;
```
Missing copyright header. It should be the same as in all other files.

NiklasGustafsson commented on src/TorchSharp/DisposeScopeManager.cs:
```diff
@@ -0,0 +1,317 @@
+            /// dispose scope, if one exists. See overloaded methods. If you wish to exclude a tensor from all sccopes,
+            /// use ExcludeGlobally.
+            /// </summary>
+            public T Exclude<T>(T exclude) where T : IDisposable
```
These overloads will probably cover 90% of all cases. Should there be a public version of the private Exclude() method that allows coverage of the rest?

mfagerlund replied:
Sounds reasonable. Fixing.

NiklasGustafsson commented on test/TorchSharpTest/NN.cs:
```diff
@@ -2591,7 +2591,6 @@ public void TestRNN2()
                 Assert.Equal(h0.shape, hN.shape);
                 Assert.Equal(new long[] { input.shape[0], input.shape[1], 20 }, output.shape);
             }
-
```
Reformatting only.

mfagerlund replied:
Fixed

NiklasGustafsson commented on test/TorchSharpTest/TestTorchTensor.cs:
```diff
@@ -16,6 +16,12 @@ namespace TorchSharp
+        public TestTensor()
+        {
+            // Avoid comma issues on international machines (Swedish, for instance...)
+            CultureInfo.DefaultThreadCurrentCulture = (CultureInfo)CultureInfo.InvariantCulture.Clone();
```
Good catch. 

I wonder if there's a need to force invariant culture on ToString() in general. I've been away from Sweden for so long, I'm afraid that I'm desensitized to number formatting, so I don't know what the right thing to do is.

mfagerlund replied:
Lots of tests failed when running a swedish install ;)

NiklasGustafsson commented on test/TorchSharpTest/TestTraining.cs:
```diff
@@ -16,7 +16,6 @@ namespace TorchSharp
 {
     public class TestTraining
     {
-
```
Superfluous reformatting change.

mfagerlund replied:
Damn, sorry for all the work this caused, I'll try to keep rider on a tighter leash from now on!

NiklasGustafsson replied:
Since I assume there are going to be massive changes in the next revision of this PR, it may be cleaner to close this one and then open up a new one. Not sure.

NiklasGustafsson commented:
@mfagerlund -- I did a review and made some comments.

mfagerlund commented:
Great, I'll try to get to them this weekend! I'm afraid Rider does a lot of
formatting changes that I didn't intend, I had to revert several files
several times because of excessive formatting changes...

/m

Den fre 19 nov. 2021 kl 18:06 skrev Niklas Gustafsson <
***@***.***>:

> @mfagerlund <https://github.com/mfagerlund> -- I did a review and made
> some comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/TorchSharp/pull/453#issuecomment-974249280>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADD556ORKNZROTXYFJT5K43UMZ73FANCNFSM5IMN6ZQQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
/Mattias Fagerlund

NiklasGustafsson commented on src/TorchSharp/DisposeScopeManager.cs:
```diff
@@ -0,0 +1,317 @@
+            public long TotalAllocatedSize =>
+                Disposables.OfType<torch.Tensor>().Sum(x => x.NumberOfElements * x.ElementSize);
+
+            public bool DontReportOnExternalDisposes { get; set; } = true;
```
I wonder about this one -- it seems to me that external disposes are the norm, and not in anyway an error. The DisposeScope should simply ignore tensors that have already been disposed.

mfagerlund replied:
I added that for some debugging purposes, it allowed me to trap when the disposes were happening. All disposes were appropriate, it turned out.

mfagerlund replied:
I removed it, if we need that kind of debugging, it's easy enough to re-introduce.

NiklasGustafsson commented on src/TorchSharp/DisposeScopeManager.cs:
```diff
@@ -0,0 +1,317 @@
+            foreach (var disposeScope in DisposeScopeStack) {
+                bool wasRemoved = disposeScope.Disposables.Remove(disposable);
+                if (wasRemoved && !disposeScope.DontReportOnExternalDisposes) {
+                    // File.AppendAllText(@"c:\slask\disp.txt",
```
Commented code -- usually a no-no in my book. With it commented, the whole if-statement is redundant. Also, see my comment on 'DontReportOnExternalDisposes'

mfagerlund replied:
Yep, it's temporary for the dispose investigation. But I'll make sure to remove it.

mfagerlund replied:
Fixed

NiklasGustafsson commented on src/TorchSharp/DisposeScopeManager.cs:
```diff
@@ -0,0 +1,317 @@
+        internal void RemoveDisposeScope(DisposeScope disposeScope)
+        {
+            lock (this) {
+                if (DisposeScopeStack.Count == 0) {
```
These tests seem like something really fundamental has been going wrong. What is a scenario where this could happen?

Should they be Debug.Assert() instead?

mfagerlund replied:
Absolutely, I'll change to that instead.

NiklasGustafsson replied:
When you edit the PR, please also start a new entry for version 0.95.4 in the RELEASENOTE.md file and describe the changes. This falls under 'API Changes'

NiklasGustafsson commented on src/TorchSharp/Tensor/Tensor.cs:
```diff
@@ -42,30 +60,35 @@ internal Tensor(IntPtr handle)
+            /// <summary>
+            /// Helper method to determine if a tensor is disposed or not.
+            /// </summary>
+            public bool IsDisposed => handle == IntPtr.Zero;
```
Linguistically speaking, this should be 'WasDisposed' but I'm not sure I can think of any precedent for using past tense in public properties.

I'm not asking for a change, I'm just curious about other APIs using anything but present tense.

mfagerlund replied:
Heh, I sometimes struggle with using Are instead of Is

lostmsu replied:
Use present tense. It is not much different from `IsEnabled` in that it tells the current state, not necessarily that some specific action took place. If you want to be really picky, rename it to `IsInvalid`.

NiklasGustafsson commented on src/TorchSharp/Tensor/Tensor.cs:
```diff
@@ -42,30 +60,35 @@ internal Tensor(IntPtr handle)
             public string? name { get; set; }
 
+            /// <summary>
+            /// Helper method to determine if a tensor is disposed or not.
```
Actually, I think calling it a 'Helper method' is underselling it. This is a bona fide state property of the tensor instance.

mfagerlund replied:
Changed to

            /// <summary>
            /// Is true if the tensor has been disposed, false otherwise.
            /// </summary>

but the entire thing is a bit self explanatory.

NiklasGustafsson commented:
By the way, please add your new test file to the 'WithCudaBinaries' project, too. They should be added as links, not by copy.

mfagerlund commented:
I added something, not sure if it's appropriate.

NuGet Version 0.95.4

*API Changes:*

Added DisposeScopeManager and torch.NewDisposeScope() to facilitate a new
solution for managing disposing of tensors with fewer usings.

Den fre 19 nov. 2021 kl 19:15 skrev Niklas Gustafsson <
***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/TorchSharp/DisposeScopeManager.cs
> <https://github.com/dotnet/TorchSharp/pull/453#discussion_r753417788>:
>
> > +            lock (Singleton ??= new DisposeScopeManager()) {
> +                return Singleton.InnerNewDisposeScope();
> +            }
> +        }
> +
> +        private DisposeScope InnerNewDisposeScope()
> +        {
> +            var disposeScope = new DisposeScope(this);
> +            DisposeScopeStack.Push(disposeScope);
> +            return disposeScope;
> +        }
> +
> +        internal void RemoveDisposeScope(DisposeScope disposeScope)
> +        {
> +            lock (this) {
> +                if (DisposeScopeStack.Count == 0) {
>
> When you edit the PR, please also start a new entry for version 0.95.4 in
> the RELEASENOTE.md file and describe the changes. This falls under 'API
> Changes'
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/TorchSharp/pull/453#discussion_r753417788>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADD556KSEAQE2UC2CX4UPQLUM2H3TANCNFSM5IMN6ZQQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
/Mattias Fagerlund

mfagerlund commented:
Starting a fresh new pull request instead!
